src/riscv/lib/src/machine_state/page_cache.rs

```diff
@@ -0,0 +1,404 @@
+// SPDX-FileCopyrightText: 2024-2025 TriliTech <contact@trili.tech>
+// SPDX-FileCopyrightText: 2025 Nomadic Labs <contact@nomadic-labs.com>
+//
+// SPDX-License-Identifier: MIT
+
+//! Per-page cache of instructions and dispatch targets.
+//!
+//! The aim is to massively speed up execution two-fold:
+//! - first, by bypassing the fetch/parse/run cycle for each instruction
+//! - secondly, by allowing hotspots of frequently re-executed sequences of instructions to be
+//!   JIT-compiled - thus running even faster.
+//!
+//! The page cache does this by populating the cache on a per-page basis. This allows the
+//! page-cache to _very strictly_ only be populated for pages that have R+X permissions set, but
+//! *are not* writable. The pleasant outcome of this is that we no longer have to keep the page
+//! cache as part of the PVM state: it is automatically always synced with data memory -- and
+//! therefore execution using the page cache is semantically identical to the fetch/parse/run
+//! cycle.
+
+use std::marker::PhantomData;
+
+use super::MachineCoreState;
+use super::ProgramCounterUpdate;
+use super::StepManyResult;
+use super::instruction::Instruction;
+use super::memory;
+use super::memory::Address;
+use super::memory::MemoryConfig;
+use crate::state_backend::ManagerBase;
+use crate::state_backend::ManagerRead;
+use crate::state_backend::ManagerReadWrite;
+use crate::traps::Exception;
+
+/// Per page, we store exactly the number of instruction halfwords we could fetch from that page's
+/// memory.
+///
+/// This is exactly half the page size - we know the instruction program counter is always
+/// halfword-aligned.
+const INSTRUCTION_ENTRIES: usize = 1
+    << memory::OFFSET_BITS
+        .checked_sub(1)
+        .expect("OFFSET_BITS is non-zero") as usize;
+
+/// Mask for indexing into a page's memory.
+const PAGE_OFFSET_MASK: u64 = (1 << memory::OFFSET_BITS) - 1;
+
+/// Instance of the page cache.
+#[expect(
+    unused,
+    reason = "PageCache is not integrated into the rest of the codebase"
+)]
+pub trait PageCache<MC: MemoryConfig, M: ManagerBase> {
```

Comment:
It's a bit hard to evaluate this trait when there is no usage/implementation of it in the PR.

Reply:
that is fair. 

In terms of usage - that's a bit hard right now, as e.g. blocked by #276. In terms of implementation, I had split out interpreted, as it lead to a significantly larger amount of code trying to land (especially in terms of required tests), but I'll shift that back into here with a second commit. Will take a little while, however